src/libraries/System.Private.CoreLib/src/System/TypedReference.cs

```diff
@@ -12,52 +13,65 @@ namespace System
             ArgumentNullException.ThrowIfNull(flds);
 
             if (flds.Length == 0)
+            {
                 throw new ArgumentException(SR.Arg_ArrayZeroError, nameof(flds));
+            }
+
+            ref byte targetRef = ref Unsafe.NullRef<byte>();
 
-            IntPtr[] fields = new IntPtr[flds.Length];
             // For proper handling of Nullable<T> don't change GetType() to something like 'IsAssignableFrom'
             // Currently we can't make a TypedReference to fields of Nullable<T>, which is fine.
             RuntimeType targetType = (RuntimeType)target.GetType();
             for (int i = 0; i < flds.Length; i++)
             {
                 RuntimeFieldInfo field = flds[i] as RuntimeFieldInfo ?? throw new ArgumentException(SR.Argument_MustBeRuntimeFieldInfo);
                 if (field.IsStatic)
+                {
                     throw new ArgumentException(SR.Format(SR.Argument_TypedReferenceInvalidField, field.Name));
+                }
 
                 if (targetType != field.GetDeclaringTypeInternal() && !targetType.IsSubclassOf(field.GetDeclaringTypeInternal()))
+                {
                     throw new MissingMemberException(SR.MissingMemberTypeRef);
+                }
 
                 RuntimeType fieldType = (RuntimeType)field.FieldType;
                 if (fieldType.IsPrimitive)
+                {
                     throw new ArgumentException(SR.Format(SR.Arg_TypeRefPrimitive, field.Name));
+                }
 
                 if (i < (flds.Length - 1) && !fieldType.IsValueType)
+                {
                     throw new MissingMemberException(SR.MissingMemberNestErr);
-
-                fields[i] = field.FieldHandle.Value;
+                }
+
+                if (i == 0)
+                {
+                    targetRef = ref GetFieldDataReference(target, field);
+                }
+                else
+                {
+                    targetRef = ref GetFieldDataReference(ref targetRef, field);
+                }
                 targetType = fieldType;
             }
 
-            TypedReference result = default;
-
-            // reference to TypedReference is banned, so have to pass result as pointer
-            unsafe
-            {
-                InternalMakeTypedReference(&result, target, fields, targetType);
-            }
-            return result;
+            return new TypedReference(ref targetRef, targetType);
         }
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        // reference to TypedReference is banned, so have to pass result as pointer
-        private static extern unsafe void InternalMakeTypedReference(void* result, object target, IntPtr[] flds, RuntimeType lastFieldType);
+        private static ref byte GetFieldDataReference(ref byte target, RuntimeFieldInfo field)
+        {
+            Debug.Assert(!Unsafe.IsNullRef(ref target));
+            int offset = GetFieldOffset(field);
```

Comment:
I think it would look better if `GetFieldDataReference(ref byte` is in the runtime-specific part next to the other `GetFieldDataReference`, and there is no `GetFieldOffset` method on TypedReference.

Reply:
> I think it would look better if GetFieldDataReference(ref byte is in the runtime-specific part next to the other GetFieldDataReference

I am genuinely surprised by this opinion. I intentionally wrote this in what I'd hoped was a runtime agnostic manner - still tracking down the Mono issues so perhaps that is wrong. I can move this, but it will currently mean a dupication of C# logic.

> and there is no GetFieldOffset method on TypedReference.

Boo. I was trying to avoid this because the `RtFieldInfo` and `RuntimeFieldInfo` split is inverted between CoreCLR and Mono. I'll see what I can do.

Reply:
> it will currently mean a dupication of C# logic.

Yes, it will mean duplicating ~2 simple lines in the implementation. I do not think an extra method is worth it to avoid duplicating ~2 simple lines, and I think it looks better (more symmetric) if the two `GetFieldDataReference` methods are next to each other. Personal style opinion.